drivers/macvlan/macvlan_setup.go

```diff
@@ -0,0 +1,214 @@
+package macvlan
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/libnetwork/osl"
+	"github.com/vishvananda/netlink"
+)
+
+// Create the netlink interface specifying the source name
+func createMacVlan(containerIfName, hostIface, macvlanMode string) (string, error) {
+	defer osl.InitOSContext()()
+
+	// Set the macvlan mode. Default is bridge mode
+	mode, err := setMacVlanMode(macvlanMode)
+	if err != nil {
+		return "", fmt.Errorf("Unsupported %s macvlan mode: %v", macvlanMode, err)
+	}
+	// verify the Docker host interface acting as the macvlan parent iface exists
+	if ok := validateHostIface(hostIface); !ok {
+		return "", fmt.Errorf("the requested host interface %s was not found on the Docker host", hostIface)
+	}
+	// Get the link for the master index (Example: the docker host eth iface)
+	hostEth, err := netlink.LinkByName(hostIface)
+	if err != nil {
+		logrus.Errorf("error occoured looking up the parent iface %s mode: %s error: %s", hostIface, macvlanMode, err)
+	}
+	// Create a macvlan link
+	macvlan := &netlink.Macvlan{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:        containerIfName,
+			ParentIndex: hostEth.Attrs().Index,
+		},
+		Mode: mode,
+	}
+	if err := netlink.LinkAdd(macvlan); err != nil {
+		// verbose but will be an issue if user creates a macvlan and ipvlan on same parent.Netlink msg is uninformative
+		logrus.Warnf("Failed to create macvlan link: %v with the error: %s", macvlan, err)
+		logrus.Warn("Ensure there are no ipvlan links using the same `--host_interface` as the macvlan iface.")
+		return "", fmt.Errorf("Failed to create the macvlan link: %v", err)
+
+	}
+	return macvlan.Attrs().Name, nil
+}
+
+// setMacVlanMode setter for one of the four macvlan port types
+func setMacVlanMode(mode string) (netlink.MacvlanMode, error) {
+	switch mode {
+	case modePrivate:
+		return netlink.MACVLAN_MODE_PRIVATE, nil
+	case modeVepa:
+		return netlink.MACVLAN_MODE_VEPA, nil
+	case modeBridge:
+		return netlink.MACVLAN_MODE_BRIDGE, nil
+	case modePassthru:
+		return netlink.MACVLAN_MODE_PASSTHRU, nil
+	default:
+		return 0, fmt.Errorf("unknown macvlan mode: %s", mode)
+	}
+}
+
+// validateHostIface check if the specified interface exists in the default namespace
+func validateHostIface(ifaceStr string) bool {
```

Comment:
Lots of methods such as validateHostIface/validateSubIface/createVlanSubIface/addVlanIface/verifyVlanDel are duplicated in ipvlan_setup.go